finalfrontier/src/idx.rs

```diff
@@ -0,0 +1,114 @@
+use std::option;
+
+/// A single index in the
+#[derive(Clone)]
+pub struct SingleIdx {
+    word_idx: u64,
+}
+
+#[derive(Clone)]
+pub struct WordWithSubwordsIdx {
+    word_idx: u64,
+    subwords: Vec<u64>,
+}
+
+impl WordWithSubwordsIdx {
+    pub(crate) fn new(word_idx: u64, subwords: impl Into<Vec<u64>>) -> Self {
+        WordWithSubwordsIdx {
+            word_idx,
+            subwords: subwords.into(),
+        }
+    }
+}
+
+/// Index representation.
+///
+/// This enum represents possible types of indices, e.g. in a vocabulary with subwords, the `Multi`
+/// variant includes all indices associated with a key.
+pub trait WordIdx: Clone {
+    /// Return the unique word index for the WordIdx.
+    fn word_idx(&self) -> u64;
+
+    /// Build a new WordIdx containing only a single index.
+    fn from_word_idx(word_idx: u64) -> Self;
+}
+
+#[allow(clippy::len_without_is_empty)]
+pub trait IndexIterator {
+    type Iter: Iterator<Item = u64>;
+
+    /// Return the number of indices.
+    fn len(&self) -> usize;
+
+    /// Return an iterator over the indices.
+    fn iter(&self) -> Self::Iter;
+}
+
+impl WordIdx for SingleIdx {
+    fn word_idx(&self) -> u64 {
+        self.word_idx
+    }
+
+    fn from_word_idx(word_idx: u64) -> Self {
+        SingleIdx { word_idx }
+    }
+}
+
+impl<'a> IndexIterator for &'a SingleIdx {
+    type Iter = option::IntoIter<u64>;
+
+    fn len(&self) -> usize {
+        1
+    }
+
+    fn iter(&self) -> Self::Iter {
+        Some(self.word_idx).into_iter()
+    }
+}
+
+impl WordIdx for WordWithSubwordsIdx {
+    fn word_idx(&self) -> u64 {
+        self.word_idx
+    }
+
+    fn from_word_idx(word_idx: u64) -> Self {
+        WordWithSubwordsIdx {
+            word_idx,
+            subwords: Vec::new(),
+        }
+    }
+}
+
+impl<'a> IndexIterator for &'a WordWithSubwordsIdx {
+    type Iter = IdxIter<'a>;
+    fn len(&self) -> usize {
+        1 + self.subwords.len()
+    }
+
+    fn iter(&self) -> Self::Iter {
+        IdxIter {
+            word_idx: Some(self.word_idx),
+            subwords: self.subwords.as_slice(),
+        }
+    }
+}
+
+/// Iterator over Indices.
+pub struct IdxIter<'a> {
+    word_idx: Option<u64>,
+    subwords: &'a [u64],
```

Comment:
Perhaps store an iterator over the slice? Simplifies the iterator implementation.